tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -2183,6 +2243,32 @@ private[optimizer] abstract class OptimizerCore(
           case _               => default
         }
 
+      case String_+ =>
+        val lhs1 = foldToStringForString_+(lhs)
+        val rhs1 = foldToStringForString_+(rhs)
+        @inline def stringDefault = BinaryOp(String_+, lhs1, rhs1)
+        (lhs1, rhs1) match {
+          case (StringLiteral(s1), StringLiteral(s2)) => 
+            StringLiteral(s1 + s2)
+          case (l, r @ StringLiteral("")) =>
+            foldBinaryOp(op, r, l)
+          case (StringLiteral(""), r) if r.tpe == StringType =>
+            r
+          case (l, BinaryOp(String_+, rl, rr)) => 
+            foldBinaryOp(String_+, BinaryOp(String_+, l, rl), rr)
+          case (l @ BinaryOp(String_+, ll, lr), StringLiteral(r)) =>
+            foldBinaryOp(String_+, ll, lr)  match {
+              case BinaryOp(String_+, llp, StringLiteral(lrp)) =>
+                BinaryOp(String_+, llp, StringLiteral(lrp + r))
+              case `l` =>
+                BinaryOp(String_+, l, StringLiteral(r))
+              case other =>
+                foldBinaryOp(String_+, other, StringLiteral(r))
+            }
+          case _ =>
+            default
```

Comment:
Since you normalize `o + ""` into `"" + o` and you normalize `(a + (b + c))` into `((a + b) + c)` (which are both good things, IMO), you can add this case too:
```scala
case (BinaryOp(String_+, StringLiteral(""), lr), _) =>
  BinaryOp(String_+, lr, rhs1)
```
which will allow to fold `(("" + a) + b)` into `(a + b)`.